rfcs/0166-nix-formatting.md

```diff
@@ -0,0 +1,1140 @@
+                           )
+                           kernel.execFormat.name;
+    }
+    ```
+- Indentation levels *must not* be "skipped", i.e. on subsequent lines, indentation can only increase by at most one level, but may decrease arbitrarily many levels.
+  - In other words: a line on indentation level 6 could be followed by a line on indentation level 1, but not the other way around.
+  - Examples:
+    ```nix
+    buildInputs = [
+        foo # <--
+      ] // lib.optionals cond [
+        bar
+      ];
+
+    attribute = { args }: let
+        foo = "bar" # <--
+      in
+        foo;
+
+    (callFunction {
+        foo = "bar"; # <--
+      }
+      arg
+    )
+
+    # This is okay, indentation increases only one level per line
+    let
+      x = {
+        a = foo
+          bar
+          baz;
+      }; # <-- The decrease by two levels here is okay though
+    in
+    null
+    ```
+
+### Expansion of expressions
+
+**Description**
+
+Unless stated otherwise, any expression that fits onto one single line will be trivially formatted as such.
+
+For sequences of items in expressions, like elements in a list, key-value pairs in attribute sets, or function arguments, the following applies:
+
+- If expanded into with multiple lines, each item should be on its own line.
+  - Grouping similar items together can be done by adding blank lines or comments between the groups instead.
+  - This also applies to the first item, so e.g. `[ firstElement` in a multi line list is not allowed.
+- Long sequences of items may be liberally expanded, even if they would fit onto one line character-wise.
+  - The motivation is to keep the information per line manageable. Usually "number of elements" is a better metric for that than "line length".
+  - The cutoff is usually determined empirically based on common usage patterns.
+
+**Examples:**
+
+```nix
+{
+  #1
+  buildInputs = [
+    foo
+    bar
+    baz
+
+    somethingElse
+  ];
+
+  #2
+  systemd.services = {
+    foo = { };
+    bar = { };
+  };
+
+  #3
+  inherit
+    lib
+    foo
+    bar
+    baz
+    ;
+}
+```
+
+### Function application
+
+**Description:**
+
+- In a function application chain, the first element is treated as the "function" and the remaining ones as "arguments".
+- The last argument receives special treatment, to better represent common coding patterns.
+- As many arguments as possible are fit onto the first line.
+  - If all but the last argument do fit, then the last argument may start on the same line.
+  - If an earlier argument does not fit onto the first line, then itself and all the following ones start on a new line. This is called the expanded form.
+  - All arguments that are not on the same line as the function are indented by one level.
+
+**Examples:**
+
+```nix
+#1
+function arg1 arg2
+
+#2
+function arg1 arg2 {
+  more = "things";
+}
+
+#3
+function arg1 arg2 arg3 # reached line limit here
+  arg4
+  arg5
+  [
+    1
+    2
+  ]
+  arg7
```

Comment:
This is a contrived example, yes, but in practice I think the current style is preferable most of the time. Especially since it is a common pattern to have `function short args (big lambda on new line) attrs` (e.g. `concatMapStringsSep`)